store/src/main/java/com/dropbox/android/external/store4/impl/RealStore.kt

```diff
@@ -179,13 +179,27 @@ internal class RealStore<Key : Any, Input : Any, Output : Any>(
             when (it) {
                 is Either.Left -> {
                     // left, that is data from network
-                    when (it.value) {
-                        is StoreResponse.Data ->
-                            // unlocking disk only if network sent data so that fresh data request
-                            // never receives disk data by mistake
-                            diskLock.complete(Unit)
-                        else ->
-                            emit(it.value.swapType())
+                    if (it.value is StoreResponse.Data || it.value is StoreResponse.NoNewData) {
+                        // unlocking disk only if network sent data or reported no new data
+                        // so that fresh data request never receives new fetcher data after
+                        // cached disk data
+                        diskLock.complete(Unit)
+                    }
+
+                    if (it.value !is StoreResponse.Data) {
+                        emit(it.value.swapType())
+                    }
+
+                    if (it.value is StoreResponse.NoNewData &&
+                        request.shouldSkipCache(CacheType.MEMORY)) {
+                        // In the special case where the request is skipping memory cache but the
+                        // fetcher returned no new data we actaully want to serve cache and SoT
+                        // data. In this case we do check the cache. If the request did not skip
+                        // memory and no new data was returned from the fetcher we do not emit
+                        // from cache again to avoid a emitting twice from the cache.
```

Comment:
This is where things get a little ugly. we give people no control over fallback but just implicitly assume cache + SoT

Reply:
Also, I think I missed the no SoT case. need to update

Reply:
this is tricky because cache does not necessarily have the latest local copy. 
also when skip cache is true if we go back to cache it is really weird. if cache is skipped, shouldn't we just fall back to SoT (especially in this code path where we are already combining w/ SoT?)